qiskit/transpiler/passes/layout/sabre_layout.py

```diff
@@ -0,0 +1,148 @@
+        """
+        from qiskit.converters import dag_to_circuit
+        from qiskit.transpiler.passes.layout.set_layout import SetLayout
+        from qiskit.transpiler.passes.layout.full_ancilla_allocation import FullAncillaAllocation
+        from qiskit.transpiler.passes.layout.enlarge_with_ancilla import EnlargeWithAncilla
+        from qiskit.transpiler.passes.layout.apply_layout import ApplyLayout
+        from qiskit.transpiler.passes.routing import SabreSwap
+        from qiskit.transpiler.passmanager import PassManager
+
+        def _layout_and_route_passmanager():
+            """Return a passmanager for a full layout and routing.
+
+            We use a factory to remove potential statefulness of passes.
+            """
+            layout_and_route = [SetLayout(initial_layout),
+                                FullAncillaAllocation(self.coupling_map),
+                                EnlargeWithAncilla(),
+                                ApplyLayout(),
+                                self.routing_pass]
+            pm = PassManager(layout_and_route)
+            return pm
+
+        def _compose_layouts(initial_layout, pass_final_layout, qregs):
+            """Return the real final_layout resulting from the composition
+            of an initial_layout with the final_layout reported by a pass.
+
+            The routing passes internally start with a trivial layout, as the
+            layout gets applied to the circuit prior to running them. So the
+            "final_layout" they report must be amended to account for the actual
+            initial_layout that was selected.
+            """
+            trivial_layout = Layout.generate_trivial_layout(*qregs)
+            pass_final_layout = Layout({trivial_layout[v.index]: p
+                                        for v, p in pass_final_layout.get_virtual_bits().items()})
+            qubit_map = Layout.combine_into_edge_map(initial_layout, trivial_layout)
+            final_layout = {v: pass_final_layout[qubit_map[v]]
+                            for v, _ in initial_layout.get_virtual_bits().items()}
+            return Layout(final_layout)
+
+        if len(dag.qubits) > self.coupling_map.size():
+            raise TranspilerError('More virtual qubits exist than physical.')
+
+        # Choose a random initial_layout.
+        if self.seed is None:
+            self.seed = np.random.randint(0, np.iinfo(np.int32).max)
+        rng = np.random.default_rng(self.seed)
+
+        physical_qubits = rng.choice(self.coupling_map.size(),
+                                     len(dag.qubits), replace=False)
+        physical_qubits = rng.permutation(physical_qubits)
+        initial_layout = Layout({q: dag.qubits[i]
+                                 for i, q in enumerate(physical_qubits)})
+
+        if self.routing_pass is None:
+            self.routing_pass = SabreSwap(self.coupling_map, 'decay')
+
+        # Do forward-backward iterations.
+        circ = dag_to_circuit(dag)
```

Comment:
This conversion might be avoid by calling the passes directly instead of using a passmanager.

Reply:
I need the circuit.reverse_ops() method which is not implemented for DAGs.